### PR TITLE
Dashboards: Refactor service to make it injectable by wire

### PR DESCRIPTION
**Pull request by spinillos**
**What this PR does / why we need it**:
We want to remove the buses from Enterprise. 

One of these buses need to create a dummy OSS version and Enterprise version and inject them in Wire. The thing is that the services where it should be injected doesn't have any dependency inversion and we need to refactor them.

**Which issue(s) this PR fixes**:

This PR refactors the whole dashboard service. The code is really old and it was used in lot of parts of the code making the PR huge 😞. 

The new parts starts from here: [dashboards folder](pkg/services/dashboards/dashboard.go) and the rest of the code are add the dependency in the other services and fix most of the tests.

**Special notes for your reviewer**:

* The dashboard's service was using an interface for the database inside other folder that I moved into the service. The implementation of this interface was inside `sqlstore` package and Wire didn't allow me to bind sqlstore to this interface. So I needed to create a new file to move all the code of the `sqlstore` to this new file.

* Most of the code it's a copy&paste from the code that we had before with a few changes to adapt it to the new code.

* I used the `mock` package from testify library to mock the behaviours. I put a [link](https://github.com/stretchr/testify#mock-package) here if you never used and avoid to make you more crazy.

* **Update 1**: I needed to move dashboards integration tests to dashboards service to fix some dependencies. The weird thing that I needed to add is to use `sqlstore.SQLStore` copy some logic from `DashboardStore` into `org_test`. I cannot use `DashboardStore` directly because it generates a dependency cycle 🙃. 

* **Update 2**: Looks like that this refactor never ends 😭. 
  

* **Update 3**: I'm still alive working on this.
   

* **Update 4**: TESTS PASSED!!!! Now fixing lot of imports hahahahaha!
   

* **Update 5**: Just removed two buses but I left another two because they need lot of extra refactor and this PR is enough big to add more. It's something that we can do after merge this colossal PR.

### CODE REVIEW

File: pkg/services/dashboards/manager/dashboard_service.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package dashboards
+package manager
```
Comment (marefr) [inline pkg/services/dashboards/manager/dashboard_service.go:1]: By convention (based on some other services) think this directory/package should be named `service`.

**Review by marefr — COMMENTED** [review]
This looks great. Awesome work. Added a minor comment.

File: pkg/services/dashboards/manager/dashboard_service.go (hunk 1 above)
Comment (spinillos) [reply pkg/services/dashboards/manager/dashboard_service.go:1]: You are very brave reviewing this PR 😄

Comment (marefr) [reply pkg/services/dashboards/manager/dashboard_service.go]: Hah, I've thought about doing this refactoring many times but never got there so very nice to see it happen and I happily help out getting it merged :pray:

File: pkg/services/dashboards/dashboard.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+type DashboardService interface {
+	SaveDashboard(ctx context.Context, dto *SaveDashboardDTO, allowUiUpdate bool) (*models.Dashboard, error)
+	ImportDashboard(ctx context.Context, dto *SaveDashboardDTO) (*models.Dashboard, error)
+	DeleteDashboard(ctx context.Context, dashboardId int64, orgId int64) error
+	MakeUserAdmin(ctx context.Context, orgID int64, userID, dashboardID int64, setViewAndEditPermissions bool) error
+	BuildSaveDashboardCommand(ctx context.Context, dto *SaveDashboardDTO, shouldValidateAlerts bool, validateProvisionedDashboard bool) (*models.SaveDashboardCommand, error)
+	UpdateDashboardACL(ctx context.Context, uid int64, items []*models.DashboardAcl) error
+}
+
+//go:generate mockery --name DashboardProvisioningService --structname FakeDashboardProvisioning --inpackage --filename dashboard_provisioning_mock.go
```
Comment (marefr) [inline pkg/services/dashboards/dashboard.go:19]: Not sure we've used the github.com/stretchr/testify/mock a lot earlier. Maybe more in Enterprise. I think we're lacking some instructions/guidance of how this works, how you generate a mock and what to do if you need to update one. Probably as its own sub-section here https://github.com/grafana/grafana/blob/main/contribute/style-guides/backend.md#testing

Can definitely be a separate task after this one is merged.

Nit. Personal taste I would add mocks to a different package, like pkg/services/dashboards/testing. Can be a potential task after this has been merged.

**Review by marefr — APPROVED** [review]
Great work. LGTM

Added a comment regarding mocking. This didn't make it for 8.4.0-beta1 and due to big changes/high risk not sure we can include this in 8.4.0 stable, but can be discussed - not including it might create problems backporting changes to 8.4.x eventually.

File: pkg/services/ngalert/tests/util.go [hunk 3]
```diff
@@ -41,7 +42,7 @@ func SetupTestEnv(t *testing.T, baseInterval time.Duration) (*ngalert.AlertNG, *
 	sqlStore := sqlstore.InitTestDB(t)
 	secretsService := secretsManager.SetupTestService(t, database.ProvideSecretsStore(sqlStore))
 	dashboardStore := databasestore.ProvideDashboardStore(sqlStore)
-	folderService := dashboardservice.ProvideFolderService(dashboardservice.ProvideDashboardService(dashboardStore), dashboardStore)
+	folderService := dashboardservice.ProvideFolderService(dashboardservice.ProvideDashboardService(dashboardStore), dashboardStore, &search.SearchService{})
```
Comment (marefr) [inline pkg/services/ngalert/tests/util.go:45]: Nit. Maybe provide nil search service to make sure it isn't used?

File: pkg/services/librarypanels/librarypanels_test.go [hunk 4]
```diff
@@ -1516,7 +1517,7 @@ func testScenario(t *testing.T, desc string, fn func(t *testing.T, sc scenarioCo
 		role := models.ROLE_ADMIN
 		sqlStore := sqlstore.InitTestDB(t)
 		dashboardStore := database.ProvideDashboardStore(sqlStore)
-		folderService := dashboardservice.ProvideFolderService(dashboardservice.ProvideDashboardService(dashboardStore), dashboardStore)
+		folderService := dashboardservice.ProvideFolderService(dashboardservice.ProvideDashboardService(dashboardStore), dashboardStore, &search.SearchService{})
```
Comment (marefr) [inline pkg/services/librarypanels/librarypanels_test.go:1520]: Nit. Maybe provide nil search service to make sure it isn't used?

File: pkg/services/libraryelements/libraryelements_test.go [hunk 5]
```diff
@@ -296,7 +297,7 @@ func testScenario(t *testing.T, desc string, fn func(t *testing.T, sc scenarioCo
 		service := LibraryElementService{
 			Cfg:           setting.NewCfg(),
 			SQLStore:      sqlStore,
-			folderService: dashboardservice.ProvideFolderService(dashboardService, dashboardStore),
+			folderService: dashboardservice.ProvideFolderService(dashboardService, dashboardStore, &search.SearchService{}),
```
Comment (marefr) [inline pkg/services/libraryelements/libraryelements_test.go:300]: Nit. Maybe provide nil search service to make sure it isn't used?

File: pkg/services/librarypanels/librarypanels_test.go [hunk 6]
```diff
@@ -1428,7 +1429,7 @@ func createFolderWithACL(t *testing.T, sqlStore *sqlstore.SQLStore, title string
 
 	dashboardStore := database.ProvideDashboardStore(sqlStore)
 	d := dashboardservice.ProvideDashboardService(dashboardStore)
-	s := dashboardservice.ProvideFolderService(d, dashboardStore)
+	s := dashboardservice.ProvideFolderService(d, dashboardStore, &search.SearchService{})
```
Comment (marefr) [inline pkg/services/librarypanels/librarypanels_test.go:1432]: Nit. Maybe provide nil search service to make sure it isn't used?

File: pkg/services/libraryelements/libraryelements_test.go [hunk 7]
```diff
@@ -207,7 +208,7 @@ func createFolderWithACL(t *testing.T, sqlStore *sqlstore.SQLStore, title string
 
 	dashboardStore := database.ProvideDashboardStore(sqlStore)
 	d := dashboardservice.ProvideDashboardService(dashboardStore)
-	s := dashboardservice.ProvideFolderService(d, dashboardStore)
+	s := dashboardservice.ProvideFolderService(d, dashboardStore, &search.SearchService{})
```
Comment (marefr) [inline pkg/services/libraryelements/libraryelements_test.go:211]: Nit. Maybe provide nil search service to make sure it isn't used?

File: pkg/services/dashboards/manager/folder_service_test.go [hunk 8]
```diff
@@ -27,6 +29,7 @@ func TestFolderService(t *testing.T) {
 		service := ProvideFolderService(
 			&dashboards.FakeDashboardService{DashboardService: ProvideDashboardService(store)},
 			store,
+			search.ProvideService(setting.NewCfg(), bus.New()),
```
Comment (marefr) [inline pkg/services/dashboards/manager/folder_service_test.go:32]: Nit. Shouldn't this be nil or mocked (if in use in test)?

**Review by marefr — APPROVED** [review]
LGTM, but added some nits. Only locked at https://github.com/grafana/grafana/pull/44588/commits/61320e4a95d69bf584c9ae922247f62596733b6b and https://github.com/grafana/grafana/pull/44588/commits/19dd9876767ed21047b1324d39854dd0243ba9d7

**Review by zserge — APPROVED** [review]

File: pkg/services/dashboards/dashboard.go (hunk 2 above)
Comment (zserge) [reply pkg/services/dashboards/dashboard.go:19]: I still can't decide on the right subpackage naming scheme here. `testing` would collide with lots of packages, including stdlib.
Naming them `mock` would make everyone who imports them come up with an alias like `dashboardmock` or `mockdashboard` or `dashboardtest` or similar.
Wouldn't it be better to name a package like that from the beginning? Akin to `iotest`, `fstest`, `httptest` and other helpers from the stdlib?

**Pull request merged**